src/js.ts

```diff
@@ -42,9 +43,7 @@ export default function analyze(source: string, visitor: Visitor) {
           case undefined:
           case "square-bracket":
           case "bracket":
-            if (!statuses.length && visitor("open-bracket", index) === false) {
-              return;
-            }
+            if (statuses.length == 0) yield [index - 1, "{"];
```

Comment:
For opening brackets, it returns the index before, and for closing brackets and filters, the index after.
I understand the reason, but I'm not sure if this can be confusing. The previous `analyze` function returns always the position after the character was found (that maybe it wasn't correct but at least consistent).

It's not a strong opinion, but I think that maybe expected result is the position before the character starts, to mimic the behavior of indexOf. For example:

```js
//for [bar] of { foo }

[4, "["]
[8, "]"]
[13, "{"]
[19, "}"]
[20, ""]
```
What do you think?

Reply:
Yeah, I also thought about this - it does feel somewhat inconsistent, but in the end I went with what I felt was easiest to use. When you're parsing a piece of JS and you're looking for opening brackets, you probably want to cut _before_ the bracket. If you are looking for a closing bracket, probably you want to cut the string _after_ the bracket. For `|>` it is a bit ambiguous, because it's more of a separator, but here too I went with ease-of-use and return the index after.

I don't have a strong opinion about it either, though. I'm fine going for consistency and always return the index after the "cause" of yielding, if you prefer. The tradeoff is having to do `index - 1` for matching most opening brackets 🤷

Reply:
I think it should return the position **before** the caracter. This is the behavior of other native functions to search in strings, like `indexOf`. For example:

```js
const code = "Hello { world }";

const position1 = code.indexOf("{"); // 6

for (const [position, cause] of iterateTopLevel(code)) {
  console.log(position, cause); // [6, "{"]
}
```

Or to be more clear:

```js
for (const [position, cause] of iterateTopLevel(code)) {
  let cause2 = code.slice(position, cause.length);
  assert(cause === cause2);
}
```